File: app/src/main/java/com/rollapp/shared/ui/group/GroupSettingsScreen.kt
package com.rollapp.shared.ui.group

import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.PickVisualMediaRequest
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.ArrowBack
import androidx.compose.material.icons.rounded.ChevronRight
import androidx.compose.material.icons.rounded.ContentCopy
import androidx.compose.material.icons.rounded.DeleteForever
import androidx.compose.material.icons.rounded.Image
import androidx.compose.material.icons.rounded.Link
import androidx.compose.material.icons.rounded.LinkOff
import androidx.compose.material.icons.rounded.Logout
import androidx.compose.material.icons.rounded.People
import androidx.compose.material.icons.rounded.Refresh
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.rollapp.shared.ui.components.InlineError
import com.rollapp.shared.ui.components.Sharing

@Composable
fun GroupSettingsScreen(
    onBack: () -> Unit,
    onOpenMembers: (String) -> Unit,
    onLeftGroup: () -> Unit,
    viewModel: GroupSettingsViewModel = hiltViewModel()
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    val context = LocalContext.current

    var renaming by remember { mutableStateOf(false) }
    var confirmLeave by remember { mutableStateOf(false) }
    var confirmDelete by remember { mutableStateOf(false) }
    var nameDraft by remember(state.group?.name) { mutableStateOf(state.group?.name.orEmpty()) }
    var descriptionDraft by remember(state.group?.description) {
        mutableStateOf(state.group?.description.orEmpty())
    }

    val coverPicker = rememberLauncherForActivityResult(
        ActivityResultContracts.PickVisualMedia()
    ) { uri -> uri?.let(viewModel::updateCover) }

    LaunchedEffect(state.finished) {
        if (state.finished) onLeftGroup()
    }

    if (renaming) {
        AlertDialog(
            onDismissRequest = { renaming = false },
            title = { Text("Edit group") },
            text = {
                Column {
                    OutlinedTextField(
                        value = nameDraft,
                        onValueChange = { nameDraft = it },
                        singleLine = true,
                        label = { Text("Group name") }
                    )
                    Spacer(Modifier.height(12.dp))
                    OutlinedTextField(
                        value = descriptionDraft,
                        onValueChange = { descriptionDraft = it },
                        label = { Text("Description") },
                        placeholder = { Text("Optional") },
                        maxLines = 3
                    )
                }
            },
            confirmButton = {
                TextButton(
                    onClick = {
                        viewModel.rename(nameDraft)
                        viewModel.updateDescription(descriptionDraft)
                        renaming = false
                    },
                    enabled = nameDraft.isNotBlank()
                ) { Text("Save") }
            },
            dismissButton = { TextButton(onClick = { renaming = false }) { Text("Cancel") } }
        )
    }

    if (confirmLeave) {
        AlertDialog(
            onDismissRequest = { confirmLeave = false },
            title = { Text("Leave this group?") },
            text = {
                Text(
                    "You'll lose access to its photos. Photos you uploaded stay with the " +
                        "group for everyone else."
                )
            },
            confirmButton = {
                TextButton(onClick = { viewModel.leaveGroup(); confirmLeave = false }) {
                    Text("Leave")
                }
            },
            dismissButton = { TextButton(onClick = { confirmLeave = false }) { Text("Cancel") } }
        )
    }

    if (confirmDelete) {
        AlertDialog(
            onDismissRequest = { confirmDelete = false },
            title = { Text("Delete \"${state.group?.name}\"?") },
            text = {
                Text(
                    "Every photo in this group is permanently deleted for all " +
                        "${state.group?.memberCount ?: 0} members. This cannot be undone."
                )
            },
            confirmButton = {
                TextButton(onClick = { viewModel.deleteGroup(); confirmDelete = false }) {
                    Text("Delete everything")
                }
            },
            dismissButton = { TextButton(onClick = { confirmDelete = false }) { Text("Cancel") } }
        )
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Group settings") },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(Icons.Rounded.ArrowBack, contentDescription = "Back")
                    }
                }
            )
        }
    ) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .verticalScroll(rememberScrollState())
        ) {
            state.error?.let { error ->
                InlineError(
                    message = error.message ?: "Something went wrong",
                    actionLabel = "Dismiss",
                    onAction = viewModel::dismissError,
                    modifier = Modifier.padding(16.dp)
                )
            }

            val group = state.group

            SettingsRow(
                icon = Icons.Rounded.People,
                title = "Members",
                subtitle = "${group?.memberCount ?: 0} in this group",
                onClick = { onOpenMembers(viewModel.groupId) }
            )

            if (group != null && group.isInviteActive) {
                SettingsRow(
                    icon = Icons.Rounded.Link,
                    title = "Invite code",
                    subtitle = group.inviteCode,
                    trailing = Icons.Rounded.ContentCopy,
                    onClick = {
                        Sharing.copyToClipboard(context, "Invite code", group.inviteCode)
                        Sharing.shareInvite(context, group.name, group.inviteCode, group.inviteLink)
                    }
                )
            }

            if (state.isAdmin) {
                HorizontalDivider(modifier = Modifier.padding(vertical = 8.dp))

                SettingsRow(
                    icon = Icons.Rounded.Image,
                    title = "Change cover photo",
                    onClick = {
                        coverPicker.launch(
                            PickVisualMediaRequest(ActivityResultContracts.PickVisualMedia.ImageOnly)
                        )
                    }
                )
                SettingsRow(
                    icon = Icons.Rounded.Refresh,
                    title = "Name and description",
                    subtitle = group?.name,
                    onClick = {
                        nameDraft = group?.name.orEmpty()
                        descriptionDraft = group?.description.orEmpty()
                        renaming = true
                    }
                )
                SettingsRow(
                    icon = Icons.Rounded.Refresh,
                    title = "Generate a new code",
                    subtitle = "The old link stops working immediately",
                    onClick = { viewModel.regenerateInvite(null) }
                )
                if (group?.isInviteActive == true) {
                    SettingsRow(
                        icon = Icons.Rounded.LinkOff,
                        title = "Turn off invites",
                        subtitle = "Nobody new can join until you make a new code",
                        onClick = viewModel::revokeInvite
                    )
                }
            }

            HorizontalDivider(modifier = Modifier.padding(vertical = 8.dp))

            SettingsRow(
                icon = Icons.Rounded.Logout,
                title = "Leave group",
                destructive = true,
                onClick = { confirmLeave = true }
            )

            if (state.isAdmin) {
                SettingsRow(
                    icon = Icons.Rounded.DeleteForever,
                    title = "Delete group",
                    subtitle = "Removes every photo for everyone",
                    destructive = true,
                    onClick = { confirmDelete = true }
                )
            }

            Spacer(Modifier.height(40.dp))
        }
    }
}

@Composable
private fun SettingsRow(
    icon: ImageVector,
    title: String,
    subtitle: String? = null,
    trailing: ImageVector = Icons.Rounded.ChevronRight,
    destructive: Boolean = false,
    onClick: () -> Unit
) {
    val tint = if (destructive) MaterialTheme.colorScheme.error else MaterialTheme.colorScheme.onSurface

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(horizontal = 20.dp, vertical = 14.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Icon(icon, contentDescription = null, tint = tint, modifier = Modifier.size(22.dp))
        Spacer(Modifier.width(18.dp))
        Column(modifier = Modifier.weight(1f)) {
            Text(text = title, style = MaterialTheme.typography.titleMedium, color = tint)
            if (!subtitle.isNullOrBlank()) {
                Text(
                    text = subtitle,
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }
        Icon(
            trailing,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.size(20.dp)
        )
    }
}

File: ios/WeWere/UI/Screens/GroupSettingsScreen.swift
import SwiftUI
import Combine
import PhotosUI

struct GroupSettingsUiState: Equatable {
    var group: Group? = nil
    var isAdmin: Bool = false
    var isSaving: Bool = false
    var error: AppError? = nil
    var finished: Bool = false
}

@MainActor
final class GroupSettingsViewModel: ObservableObject {
    @Published private(set) var state = GroupSettingsUiState()

    let groupId: String
    private let groupRepository: GroupRepository
    private let saving = CurrentValueSubject<Bool, Never>(false)
    private let errorFlow = CurrentValueSubject<AppError?, Never>(nil)
    private let finished = CurrentValueSubject<Bool, Never>(false)
    private var cancellables = Set<AnyCancellable>()

    init(groupId: String, container: AppContainer = .shared) {
        self.groupId = groupId
        groupRepository = container.groupRepository

        groupRepository.observeGroup(groupId: groupId)
            .combineLatest(groupRepository.observeMembership(groupId: groupId), saving, errorFlow)
            .combineLatest(finished)
            .map { first, done in
                let (group, membership, isSaving, error) = first
                return GroupSettingsUiState(group: group, isAdmin: membership?.isAdmin == true,
                                            isSaving: isSaving, error: error, finished: done)
            }
            .receive(on: DispatchQueue.main)
            .sink { [weak self] in self?.state = $0 }
            .store(in: &cancellables)
    }

    func rename(_ name: String) { run { await self.groupRepository.updateGroup(groupId: self.groupId, name: name, description: nil).unit } }
    func updateDescription(_ description: String) { run { await self.groupRepository.updateGroup(groupId: self.groupId, name: nil, description: description).unit } }
    func updateCover(_ url: URL) {
        run {
            let outcome = await self.groupRepository.updateCoverPhoto(groupId: self.groupId, url: url).unit
            try? FileManager.default.removeItem(at: url)
            return outcome
        }
    }
    func regenerateInvite(_ expiresAt: Millis?) { run { await self.groupRepository.regenerateInviteCode(groupId: self.groupId, expiresAt: expiresAt).unit } }
    func revokeInvite() { run { await self.groupRepository.revokeInvite(groupId: self.groupId) } }
    func leaveGroup() { run(finishOnSuccess: true) { await self.groupRepository.leaveGroup(groupId: self.groupId) } }
    func deleteGroup() { run(finishOnSuccess: true) { await self.groupRepository.deleteGroup(groupId: self.groupId) } }

    func dismissError() { errorFlow.send(nil) }

    private func run(finishOnSuccess: Bool = false, _ block: @escaping () async -> Outcome<Void>) {
        if saving.value { return }
        Task {
            saving.send(true)
            errorFlow.send(nil)
            switch await block() {
            case .success: if finishOnSuccess { finished.send(true) }
            case .failure(let error): errorFlow.send(error)
            }
            saving.send(false)
        }
    }
}

/// Drafts for the rename dialog, shared between its fields and its Save button.
final class GroupDraft: ObservableObject {
    @Published var name: String
    @Published var description: String
    init(name: String, description: String) { self.name = name; self.description = description }
}

struct GroupSettingsScreen: View {
    let onBack: () -> Void
    let onOpenMembers: (String) -> Void
    let onLeftGroup: () -> Void

    @StateObject private var viewModel: GroupSettingsViewModel
    @EnvironmentObject private var overlays: Overlays
    @State private var coverPick: PhotosPickerItem? = nil

    init(groupId: String, onBack: @escaping () -> Void, onOpenMembers: @escaping (String) -> Void, onLeftGroup: @escaping () -> Void) {
        self.onBack = onBack
        self.onOpenMembers = onOpenMembers
        self.onLeftGroup = onLeftGroup
        _viewModel = StateObject(wrappedValue: GroupSettingsViewModel(groupId: groupId))
    }

    var body: some View {
        let state = viewModel.state
        let group = state.group
        VStack(spacing: 0) {
            RollTopBar(title: "Roll settings", onBack: onBack)
            ScrollView {
                VStack(spacing: 0) {
                    if let error = state.error {
                        InlineError(message: error.message ?? "Something went wrong",
                                    actionLabel: "Dismiss", onAction: viewModel.dismissError)
                            .padding(16)
                    }

                    SettingsRow(icon: RollIcon.people, title: "Members",
                                subtitle: "\(group?.memberCount ?? 0) in this roll") {
                        onOpenMembers(viewModel.groupId)
                    }

                    if let group, group.isInviteActive {
                        SettingsRow(icon: RollIcon.link, title: "Invite code", subtitle: group.inviteCode, trailing: RollIcon.copy) {
                            Sharing.copyToClipboard(group.inviteCode)
                            Sharing.shareInvite(groupName: group.name, code: group.inviteCode, link: group.inviteLink)
                        }
                    }

                    if state.isAdmin {
                        Divider().overlay(OutlineVariant).padding(.vertical, 8)

                        PhotosPicker(selection: $coverPick, matching: .images) {
                            SettingsRowLabel(icon: RollIcon.image, title: "Change cover photo")
                        }
                        .buttonStyle(PlainPressStyle())
                        .onChange(of: coverPick) { _, item in
                            guard let item else { return }
                            Task {
                                let urls = await PhotoPicking.stage([item], into: AppContainer.shared.capturesDir)
                                coverPick = nil
                                if let url = urls.first { viewModel.updateCover(url) }
                            }
                        }
                        SettingsRow(icon: RollIcon.refresh, title: "Name and description", subtitle: group?.name) {
                            editDetails(group)
                        }
                        SettingsRow(icon: RollIcon.refresh, title: "Generate a new code",
                                    subtitle: "The old link stops working immediately") {
                            viewModel.regenerateInvite(nil)
                        }
                        if group?.isInviteActive == true {
                            SettingsRow(icon: RollIcon.linkOff, title: "Turn off invites",
                                        subtitle: "Nobody new can join until you make a new code") {
                                viewModel.revokeInvite()
                            }
                        }
                    }

                    Divider().overlay(OutlineVariant).padding(.vertical, 8)

                    SettingsRow(icon: RollIcon.logout, title: "Leave roll", destructive: true) {
                        overlays.confirm(
                            title: "Leave this roll?",
                            message: "You'll lose access to its photos. Photos you uploaded stay with the roll for everyone else.",
                            confirmLabel: "Leave"
                        ) { viewModel.leaveGroup() }
                    }

                    if state.isAdmin {
                        SettingsRow(icon: RollIcon.deleteForever, title: "Delete roll",
                                    subtitle: "Removes every photo for everyone", destructive: true) {
                            overlays.confirm(
                                title: "Delete \"\(group?.name ?? "")\"?",
                                message: "Every photo in this roll is permanently deleted for all \(group?.memberCount ?? 0) members. This cannot be undone.",
                                confirmLabel: "Delete everything"
                            ) { viewModel.deleteGroup() }
                        }
                    }

                    Spacer().frame(height: 40)
                }
            }
        }
        .background(Ink.ignoresSafeArea())
        .onChange(of: state.finished) { _, finished in if finished { onLeftGroup() } }
    }

    private func editDetails(_ group: Group?) {
        let draft = GroupDraft(name: group?.name ?? "", description: group?.description ?? "")
        overlays.show(DialogSpec(
            title: "Edit roll",
            content: AnyView(GroupDraftFields(draft: draft)),
            confirm: .init(label: "Save", enabled: { !draft.name.trimmingCharacters(in: .whitespaces).isEmpty }) {
                overlays.closeDialog()
                viewModel.rename(draft.name)
                viewModel.updateDescription(draft.description)
            },
            dismiss: .init(label: "Cancel") { overlays.closeDialog() }
        ))
    }
}

private struct GroupDraftFields: View {
    @ObservedObject var draft: GroupDraft

    var body: some View {
        VStack(spacing: 12) {
            RollTextField(label: "Roll name", text: $draft.name, capitalization: .words,
                          cornerRadius: RollShapes.medium, chipBackground: Raised)
            RollTextField(label: "Description", text: $draft.description, placeholder: "Optional",
                          singleLine: false, minLines: 1, maxLines: 3,
                          cornerRadius: RollShapes.medium, chipBackground: Raised)
        }
    }
}

struct SettingsRowLabel: View {
    let icon: String
    let title: String
    var subtitle: String? = nil
    var trailing: String = RollIcon.chevronRight
    var destructive: Bool = false

    var body: some View {
        let tint = destructive ? ErrorColor : Ivory
        HStack(spacing: 0) {
            Image(systemName: icon).font(.system(size: 19, weight: .medium)).foregroundStyle(tint)
                .frame(width: 22)
            Spacer().frame(width: 18)
            VStack(alignment: .leading, spacing: 2) {
                Text(title).rollStyle(RollType.titleMedium, color: tint)
                if let subtitle, !subtitle.isEmpty {
                    Text(subtitle).rollStyle(RollType.bodyMedium, color: IvoryMuted)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            Image(systemName: trailing).font(.system(size: 16, weight: .semibold)).foregroundStyle(IvoryMuted)
        }
        .padding(.horizontal, 20).padding(.vertical, 14)
        .contentShape(Rectangle())
    }
}

struct SettingsRow: View {
    let icon: String
    let title: String
    var subtitle: String? = nil
    var trailing: String = RollIcon.chevronRight
    var destructive: Bool = false
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            SettingsRowLabel(icon: icon, title: title, subtitle: subtitle, trailing: trailing, destructive: destructive)
        }
        .buttonStyle(PlainPressStyle())
    }
}
